packages/daimo-api/src/utils/cache.ts

```diff
@@ -0,0 +1,74 @@
+import { assert } from "@daimo/common";
+
+/**
+ * Lazily executes `fn` and caches the result ensuring only one execution at a time.
+ *
+ * `cacheTime` is duration in ms after cached value is considred expired,
+ * and must be refreshed before it can be returned.
+ *
+ * `staleTime` is an optional duration that refreshes
+ * the cache before it expires. Use it to avoid blocking after cache expires.
+ * Must be smaller than `cacheTime`.
+ *
+ */
+export function lazyCache<T>(
```

Comment:
this is interesting. feels like a lot of complexity for just `getFinalizedBlock`

finalizedBlock changes slowly & doesn't affect UI latency. is the goal to reduce JSON RPC API calls?

Reply:
Finalized block is done on every account history fetch. It was [second slowest operation](https://github.com/daimo-eth/daimo/issues/1180#issuecomment-2186485047) in `getAccountHistory`. Now we share finalized block data across all users limiting RPC calls from variable `n` to constant `2` every ~2 sec, where `n` is number of connected users.